esphome/components/esp32/gpio.py

```diff
@@ -59,65 +72,61 @@ def _translate_pin(value):
     return _lookup_pin(value)
 
 
-_ESP_SDIO_PINS = {
-    6: "Flash Clock",
-    7: "Flash Data 0",
-    8: "Flash Data 1",
-    11: "Flash Command",
+class Esp32ValidationFunctions:
+    def __init__(self, pin_validation, usage_validation):
+        self.pin_validation = pin_validation
+        self.usage_validation = usage_validation
+
+
+_esp32_validations = {
+    VARIANT_ESP32: Esp32ValidationFunctions(
+        pin_validation=esp32_validate_gpio_pin, usage_validation=esp32_validate_supports
+    ),
+    VARIANT_ESP32S2: Esp32ValidationFunctions(
+        pin_validation=esp32_s2_validate_gpio_pin,
+        usage_validation=esp32_s2_validate_supports,
+    ),
+    VARIANT_ESP32C3: Esp32ValidationFunctions(
+        pin_validation=esp32_c3_validate_gpio_pin,
+        usage_validation=esp32_c3_validate_supports,
+    ),
+    VARIANT_ESP32S3: Esp32ValidationFunctions(
+        pin_validation=esp32_s3_validate_gpio_pin,
+        usage_validation=esp32_s3_validate_supports,
+    ),
+    VARIANT_ESP32H2: Esp32ValidationFunctions(
+        pin_validation=esp32_h2_validate_gpio_pin,
+        usage_validation=esp32_h2_validate_supports,
+    ),
 }
 
 
 def validate_gpio_pin(value):
     value = _translate_pin(value)
-    if value < 0 or value > 39:
-        raise cv.Invalid(f"Invalid pin number: {value} (must be 0-39)")
-    if value in _ESP_SDIO_PINS:
-        raise cv.Invalid(
-            f"This pin cannot be used on ESP32s and is already used by the flash interface (function: {_ESP_SDIO_PINS[value]})"
-        )
-    if 9 <= value <= 10:
-        _LOGGER.warning(
-            "Pin %s (9-10) might already be used by the "
-            "flash interface in QUAD IO flash mode.",
-            value,
-        )
-    if value in (20, 24, 28, 29, 30, 31):
-        # These pins are not exposed in GPIO mux (reason unknown)
-        # but they're missing from IO_MUX list in datasheet
-        raise cv.Invalid(f"The pin GPIO{value} is not usable on ESP32s.")
-    return value
+    variant = CORE.data[KEY_ESP32][KEY_VARIANT]
+    if variant not in VARIANTS:
```

Comment:
```suggestion
    if variant not in _esp32_validations:
```

If that's what we're using, then we should check against it.